Add development, review, and merge processes (#1210)
conda/infrastructure: HOW_WE_USE_GITHUB.md

### How should we approach feature development?

For new features, begin by creating documentation for the new feature within
the project so that you can get early input on the design. Documentation
includes design mockups, API reference docs, recording the planned approach
in the feature issue, and creating CLI argument demos without the
functionality of the feature.

### What is our change process?

For larger features, break down the work into smaller, manageable issues
that are added to the backlog. As long as a feature remains on the roadmap
or backlog, do not create long-lived feature branches that span multiple
pull requests. Instead, you should integrate small slices of an overall
feature directly into the main branch to avoid complex integration challenges.

### Should we make unrelated changes at the same time?

When making changes, try to follow the Campsite Rule to leave things better
than when you found them. You should enhance the code you encounter, even if
primary goal is unrelated. This could involve refactoring small sections,
improving readability, or fixing minor bugs.

## Code Review and Merging

### What are the review requirements?

#### Standard Review

> Most code changes require one reviewer from the conda-maintainers team.

Author: Hey @jezdez, good catch, I updated it to say someone on the maintainer team for that repository.

In order to not have to manually type or copy/paste the above repeatedly, note that it's possible to add text for the most commonly-used responses via [GitHub's "Add Saved Reply" option][docs-saved-reply].

## Development Processes

> 

Author: Thanks for the feedback, I added this section!

### How should we approach feature development?

For new features, begin by creating documentation for the new feature within
the project so that you can get early input on the design. Documentation
includes design mockups, API reference docs, recording the planned approach
in the feature issue, and creating CLI argument demos without the

> functionality of the feature.

Author: I updated this to say we need a feature issue first. I also renamed Documentation to a Specification, hopefully that deconflicts it with things in the docs folder.

### Should we make unrelated changes at the same time?

When making changes, try to follow the Campsite Rule to leave things better
than when you found them. You should enhance the code you encounter, even if
primary goal is unrelated. This could involve refactoring small sections,
improving readability, or fixing minor bugs.

## Code Review and Merging

### What are the review requirements?

#### Standard Review

Most code changes require one reviewer from someone on the maintainer team for
the repository. Instead of waiting for someone on the team to review it,
directly requesting a review from the person you previously identified to work
with is preferred to optimize teamwork. If you paired with them during
development, continuous review counts as this requirement.

#### Second Review

Required only when the code author or the first reviewer feels like it is
necessary to get another set of eyes on a proposed change. In this case, they
add someone specific through GitHub's Request Review feature with a comment on
what they want the person to look for.

### What are the code review best practices?

If you are conducting a review, adhere to these best practices:

> - Provide comprehensive feedback in the first review to minimize review rounds

Reviewer: ```suggestion
- Provide comprehensive feedback in the first review to minimize review rounds
- For core reviews, do not simply "comment" on the review; either submit a review as
   "approve" or "request changes" to clearly communicate to the author what is needed
```

This is a practice that seems like would help during our reviews. Happy to discuss here if anyone disagrees though.

Author: I really like that we are discussing this and can make this practice more consistent on when we do certain types of review selections.

The other problem with Request Changes is someone else can't approve of the code with out clearing the review. Clearing a review requires entered justification which makes it feel like you are breaking the rules. This slows down cadence unless the first reviewer is ready to re-review quickly.

I'm thinking request changes should be reserved for things that would be harmful to the code base if merged as is. I added another suggestion to capture this.

- Follow-up reviews should focus on whether requested changes resolve original
  comments
- Code should be production-ready and maintainable when merged, but doesn't
  need to be perfect
- If providing feedback outside the core review focus (nitpicks, tips,
  suggestions), clearly mark these as non-blocking comments that don't need to
  be addressed before merging.

### How do we merge code?

If you are the approving reviewer (typically the first reviewer, or the second
reviewer when needed) and you have completed your review and approved the
changes, you should merge the code immediately to maintain development
velocity.

Normally, we use squash and merge to keep a clean git history. If you are

> merging a Pull Request, help ensure that the Pull Request title is updated.

Reviewer: ```suggestion
merging a pull request, help ensure that the pull request title is updated.
```

A "pull request" is not a proper noun in this context. 🤓

Author: Fixed, thanks!

Also you should update the description to remove the list of commit messages

> and instead add a detailed explanation of the "what" and "why" of changes.

Reviewer: @danyeaw,

I get where you're coming from here, but I think we need a little more direction to make sure these messages are consistently formatted. I personally never do this, so I think it would be helpful to have a template or some form of automation to enforce what these messages look like. Otherwise, everyone will interpret this in their own special way.

Reviewer: While that is a lot to dig through, whenever I'm looking at a git blame and trying to figure out why a change was introduced, it's super easy to go back to the original PR and the issue which is where I really look to find out why a change was made.

Furthermore, the release notes are also incredibly useful for this too. It's actually pretty rare if ever that I find myself examine the commit history.

@danyeaw, how would you simplify that squash message? Would you just want a link to the original pull request? Trying to summarize all of that text manually seems like a headache.

Author: The link to the original PR is in the title of the squashed commit already which is good.

I noticed that they added a feature in 2022 to customize this: https://github.blog/changelog/2022-08-23-new-options-for-controlling-the-default-commit-message-when-merging-a-pull-request/

We could change the option from what it is today:
Default message: commit title and commit message (if the pull request contains a single commit), or the pull request title and number and list of commits (if the pull request contains multiple commits)

to:
Pull request title and description: pull request title and number on the first line; pull request description starting on the third line

This would then let the PR author put more details about the how in the PR description. If someone wants to go look through the individual commits, they could then open the PR to do so. I don't think the list of commits in the message right now is very useful.

Author: Thanks, since this isn't a standard yet, let's enable it for one of the repos like conda-pypi to try it out. Once we have some feedback about it, we can update the guidance. For now, I removed the sentence about updating the description.

In order to not have to manually type or copy/paste the above repeatedly, note that it's possible to add text for the most commonly-used responses via [GitHub's "Add Saved Reply" option][docs-saved-reply].

## Development Processes

The following are practices the conda organization encourages for feature
development. While we recommend projects under the conda organization adopt
these practices, they are not strictly required.

### How should we approach feature development?

> For new features, first open an issue if one doesn’t exist. Once the feature
> has been accepted by the issue's status transitioning from the Sorting state to
> Refinement, create a specification to gather early feedback. This can include
> mockups, API/command references, a written plan in the issue, and sample CLI
> arguments (without functionality).

Reviewer 2: ```suggestion
For new features, first open an issue if one doesn’t exist. Once the feature request
has been accepted (indicated by the issue's status transitioning from "Sorting to
"Refinement"), create a specification to gather early feedback. This can include
mockups, API/command references, a written plan in the issue, and sample CLI
arguments (without functionality).
```

primary goal is unrelated. This could involve refactoring small sections,
improving readability, or fixing minor bugs.

## Code Review and Merging

### What are the review requirements?

#### Standard Review

Most code changes require one reviewer from someone on the maintainer team for
the repository. Instead of waiting for someone on the team to review it,
directly requesting a review from the person you previously identified to work
with is preferred to optimize teamwork. If you paired with them during
development, continuous review counts as this requirement.

#### Second Review

Required only when the code author or the first reviewer feels like it is
necessary to get another set of eyes on a proposed change. In this case, they
add someone specific through GitHub's Request Review feature with a comment on
what they want the person to look for.

### What are the code review best practices?

If you are conducting a review, adhere to these best practices:

- Provide comprehensive feedback in the first review to minimize review rounds
- Reserve Request Changes for blocking issues (bugs or other major problems) —
  Select Comment for suggestions and improvements

> - Reserve Request Changes for blocking issues (bugs or other major problems) —
>   Select Comment for suggestions and improvements

Reviewer: ```suggestion
```
Looks like this line somehow got duplicated.

In order to not have to manually type or copy/paste the above repeatedly, note that it's possible to add text for the most commonly-used responses via [GitHub's "Add Saved Reply" option][docs-saved-reply].

## Development Processes

The following are best practices the conda organization encourages for feature
development. While we recommend projects under the conda organization adopt
these practices, they are not strictly required.

### How should we approach feature development?

> For new features, begin by creating an issue for the feature if one doesn't
> already exist. Once a new feature has been added to the backlog, next create a
> specification for the new feature within the project so that you can get early
> input on the design. Specification includes design mockups, API or command
> reference docs, recording the planned approach in the feature issue, and
> creating CLI argument demos without the functionality of the feature.

Reviewer 2: Let's make it a little simpler. 

```suggestion
For new features, first open an issue if one doesn’t exist. After it’s added to the backlog, create a specification to gather early feedback. This can include mockups, API/command references, a written plan in the issue, and sample CLI arguments (without functionality).
```

Author: Hi @ForgottenProgramme, thanks, I incorporated this suggestion!
